src/core/lib/security/credentials/ssl/ssl_credentials.cc

```diff
@@ -66,12 +68,61 @@ grpc_ssl_credentials::create_security_connector(
   absl::optional<std::string> overridden_target_name =
       args->GetOwnedString(GRPC_SSL_TARGET_NAME_OVERRIDE_ARG);
   auto* ssl_session_cache = args->GetObject<tsi::SslSessionLRUCache>();
-  grpc_core::RefCountedPtr<grpc_channel_security_connector> sc =
-      grpc_ssl_channel_security_connector_create(
-          this->Ref(), std::move(call_creds), &config_, target,
-          overridden_target_name.has_value() ? overridden_target_name->c_str()
-                                             : nullptr,
-          ssl_session_cache == nullptr ? nullptr : ssl_session_cache->c_ptr());
+  tsi_ssl_session_cache* tsi_ssl_session_cache =
+      ssl_session_cache == nullptr ? nullptr : ssl_session_cache->c_ptr();
+
+  const tsi_ssl_root_certs_store* root_store;
+  // Use default (e.g. OS) root certificates if the user did not pass any root
+  // certificates.
+  if (config_.pem_root_certs == nullptr) {
+    const char* pem_root_certs =
+        grpc_core::DefaultSslRootStore::GetPemRootCerts();
+    if (pem_root_certs == nullptr) {
+      gpr_log(GPR_ERROR, "Could not get default pem root certs.");
+      return nullptr;
+    }
+    size_t root_len = strlen(pem_root_certs);
+    char* default_roots = strcpy(new char[root_len + 1], pem_root_certs);
+    config_.pem_root_certs = default_roots;
+    root_store = grpc_core::DefaultSslRootStore::GetRootStore();
+  } else {
+    config_.pem_root_certs = config_.pem_root_certs;
+    root_store = nullptr;
+  }
+
+  grpc_core::RefCountedPtr<grpc_channel_security_connector> sc = nullptr;
+  // We need a separate factory and ctx if there's a cache in the channel
+  // args
+  if (tsi_ssl_session_cache != nullptr) {
+    // Get a different client handshaker factory and give to it the security
+    // connector
+    tsi_ssl_client_handshaker_factory* factory_with_cache = nullptr;
+    grpc_security_status status = initialize_client_handshaker_factory(
+        &config_, config_.pem_root_certs, root_store, tsi_ssl_session_cache,
+        &factory_with_cache);
+    if (status != GRPC_SECURITY_OK) {
+      return nullptr;
+    }
+    sc = grpc_ssl_channel_security_connector_create(
+        this->Ref(), std::move(call_creds), &config_, target,
+        overridden_target_name.has_value() ? overridden_target_name->c_str()
+                                           : nullptr,
+        factory_with_cache);
+    tsi_ssl_client_handshaker_factory_unref(factory_with_cache);
+  } else {
+    grpc_security_status status = initialize_client_handshaker_factory(
```

Comment:
Can we move this call to `initialize_client_handshaker_factory` to the constructor?

You can save the `status` as a member field if you want, and check the status here.

Otherwise, we're not getting the performance gain I think.

Reply:
I think we still get the efficiency gain because `initialize_client_handshaker_factory` just returns the class `client_handshaker_factory_` if it already exists, so the credentials really just initiate it once the first time this is called.

Doing it in the constructor, storing the status, then checking the status at this location would just move the heavy lifting to construction time of the credentials from the initial establishing of the connection - all of this happens before the first connection regardless, so I don't think shifting it necessarily gets any notable gain, just displaces it.

WDYT?

Reply:
We will get some efficiency gain, but the current code is not thread-safe: if you make multiple concurrent calls to `create_security_connector`, then there will be race conditions over populating the handshaker factory. To make it thread-safe, we would need to add a lock to guard the code block that populates and fetches the handshaker factory. However, I strongly suspect that contention over this lock would eat away at most of the performance gains that we want (at in the use cases where you're continuously creating a large volume of connections).

If we move everything to the constructor, then the cached handshaker factory is constant throughout the lifetime of the `grpc_ssl_credentials` object, and we avoid the lock contention issue.